src/Adapter/FilesystemItems.php

```diff
@@ -0,0 +1,209 @@
+        }
+
+        if (null === $version) {
+            $this->filesystem->remove("{$this->path}/{$id}");
+
+            return;
+        }
+
+        try {
+            $highest = $this->get($id)->getVersion();
+        } catch (ItemNotFound $e) {
+            return;
+        }
+
+        if ($version->toInt() < $highest->toInt()) {
+            throw new UnexpectedVersionNumber($id, $version, $highest);
+        }
+
+        if (1 === $version->toInt()) {
+            $this->filesystem->remove("{$this->path}/{$id}");
+
+            return;
+        }
+
+        $this->filesystem->remove("{$this->path}/{$id}/{$version->toInt()}.xml");
+    }
+
+    public function get(ItemId $id, ?ItemVersionNumber $version = null) : ItemVersion
+    {
+        if (!is_dir("{$this->path}/{$id}")) {
+            throw new ItemNotFound($id);
+        }
+
+        if (null === $version) {
+            $versions = iterator_to_array($this->getVersions($id));
+            rsort($versions);
+            $version = $versions[0];
+        }
+
+        if (!$content = @fopen($file = "{$this->path}/{$id}/{$version->toInt()}.xml", 'rb')) {
+            throw new VersionNotFound($id, $version);
+        }
+
+        if (!$hash = md5_file($file)) {
+            throw new RuntimeException("Failed to hash {$file}");
+        }
+
+        return new ItemVersion($id, $version, $content, $hash);
+    }
+
+    public function list(int $limit = 10, ?ItemId $startAt = null) : ItemListPage
+    {
+        $ids = $this->getIds();
+
+        if (null === $startAt) {
+            $ids = new LimitIterator($ids, 0, $limit + 1);
+        } else {
+            $found = null;
+            $i = 0;
+            foreach ($ids as $id) {
+                if ($id == $startAt) {
+                    $found = $i;
+                    break;
+                }
+                $i++;
+            }
+
+            if (null === $found) {
+                $ids = new EmptyIterator();
+            } else {
+                $ids = new LimitIterator($ids, $found, $limit + 1);
+            }
+        }
+
+        $ids = array_values(iterator_to_array($ids));
+
+        if (count($ids) > $limit) {
+            $nextId = array_pop($ids);
+        }
+
+        return new ItemListPage($ids, $nextId ?? null);
+    }
+
+    public function count() : int
+    {
+        return iterator_count($this->getIds());
+    }
+
+    public function getIterator() : Traversable
+    {
+        $items = new MapIterator(
+            $this->getIds(),
+            function (ItemId $id) : ItemVersion {
+                return $this->get($id);
+            }
+        );
+
+        foreach ($items as $item) {
+            yield $item;
+        }
+    }
+
+    private function getIds() : Iterator
+    {
+        return new MapIterator(
+            new CallbackFilterIterator(
+                new FilesystemIterator("{$this->path}/"),
+                function (SplFileInfo $file) : bool {
+                    return $file->isDir();
```

Comment:
may also help to skip hidden directories like `.git` (and in general `.something`)

Reply:
Hmm, both are valid IDs. 🤔

Reply:
b859977224979326b7d73c4091ccde48048192ac

Reply:
As long as it's clear no other subdirectories are allowed everything is consistent. Meant someone may want to point this to a Git repository, would have to use a subfolder in that repository (good practice anyway)

Reply:
Trying to determine if the contents should be allowed to be externally written (also affects https://github.com/libero/content-api-bundle/pull/6#discussion_r229990036). If so, they need sensible handling of broken data structures. If not, then the dummy API (and similar) would need to import content through some mechanism (CLI?) rather than having them dropped in a folder.